### PR TITLE
Test: Disable browser cache through selenium #8436

### LINKED ISSUE #8436: Testing: Disable cache in selenium
**Description of feature/enhancement**

Disable cache in selenium.

**Justification**
Currently cache is not disabled which causes some tests to fail and when used in conjunction with godmode can cause godmode to work unexpectedly. This creates additional gotchas if new developers forget to disable it. Disabling it within selenium can smoothen this process.

### PR DESCRIPTION
**Pull request by Crphang**
Fixes #8436 

1. Set chrome browser to incognito mode. 
2. Set firefox `browser.cache.disk.enable` to `False`.

### CODE REVIEW

**Comment by Crphang** [comment]
Ready for review!

**Comment by whipermr5** [comment]
Can update the instructions about disabling cache in `development.md` since it's now automated. Maybe change it to a note saying that cache is automatically disabled during automated testing.

**Comment by Crphang** [comment]
Change on docs added

File: docs/development.md [hunk 1]
```diff
@@ -142,21 +144,13 @@ It is recommended to use Firefox 46.0 as this is the browser used in CI build (T
 
 * If you want to use a Firefox version other than your computer's default, specify the custom path in `test.firefox.path` value in `test.properties`.
 
-* If you are planning to test changes to JavaScript code, disable JavaScript caching for Firefox:
-  * Enter `about:config` into the Firefox address bar and set `network.http.use-cache` (or `browser.cache.disk.enable` in newer versions of Firefox) to `false`.
-
```
Comment (whipermr5) [inline docs/development.md]: On second thoughts, can leave the two paragraphs in. I realised they are about manual (not automated) testing 😅

File: docs/development.md [hunk 2]
```diff
@@ -133,6 +133,8 @@ It is recommended to use Firefox 46.0 as this is the browser used in CI build (T
 **NOTE**
 > The dev server sets its time zone to UTC at startup.
 
+> Selenium turns off browser cache for both Firefox and Chrome
```
Comment (whipermr5) [inline docs/development.md]: Add a full stop.

**Comment by LiHaoTan** [comment]
I'm just wondering if we even need to do this, it appears WebDriver always uses a new temporary profile so browser cache is never preserved. The cache we are seeing may be due to one of the following:

- `.jsp`, `.tag` not propagated to the server
- `.jsp`, `.tag` not built
- `.js` not built


> By default, geckodriver will create a new profile in this location.

https://github.com/mozilla/geckodriver/README.md

> By default, ChromeDriver will create a new temporary profile for each session.

https://sites.google.com/a/chromium.org/chromedriver/capabilities

**Comment by Crphang** [comment]
Will investigate!

**Comment by wkurniawan07** [comment]
> By default, geckodriver will create a new profile in this location. https://github.com/mozilla/geckodriver/blob/master/README.md

@LiHaoTan the fact that you're referring to geckodriver might very well mean that it's not applicable to us yet until we move to Selenium 3.

**Comment by LiHaoTan** [comment]
@wkurniawan07 The reason I'm referring to that is because the documentation before for `WebDriver` or even `FirefoxDriver` is hard to find. It may be listed in the specification though I did not check but I have verified the behavior for `FirefoxDriver` is the same.

**Comment by wkurniawan07** [comment]
Actually, @Crphang, do you open this issue because you have experienced the behaviour yourself or it is just something that you think is possible but have yet to verify?
I personally have never encountered such behaviour but it's likely because I don't use Firefox other than for testing this project.

**Comment by Crphang** [comment]
@wkurniawan07 I encountered this behaviour when using godmode, when there are unnecessary files that are added which should not be affected.

**Comment by wkurniawan07** [comment]
@Crphang just checking, it's not because you forget to do `npm run build` right?

**Comment by Crphang** [comment]
@wkurniawan07 Yup it's not, I usually do the test after I do manual testing. The changes is unrelated to the javascript that was changed, usually some other tests and some other HTML pages, as if it's due to the sequence in which the HTML was run. I will try to replicate it.

**Comment by Crphang** [comment]
Referring to PR: https://github.com/TEAMMATES/teammates/pull/8373.

Running godmode with cache on gave unintended results like the highlighted file:

files_changed

and these html changes is unintended:

should not be there

I turned off cache and re-ran test twice again with godmode and the change is un-done. Anyone else  experienced similar results with cache on?

**Comment by whipermr5** [comment]
Could it be that these unintended changes were not due to cached javascript, but simply a result of running unstable tests with GodMode enabled? Perhaps run the tests multiple times to see if the changes are reintroduced?

**Comment by Crphang** [comment]
@whipermr5 @LiHaoTan, after several tries, it does seem like whether browser cache is on or not does to help increase the stability of tests. There are other problems that causes godmode and tests to work unexpectedly which requires more effort on. Closing this PR and corresponding for now until we know for certain that turning off cache will be helpful in a specific known area.

**Pull request closed without merging**